torch/csrc/jit/python/init.cpp

```diff
@@ -897,13 +898,65 @@ void initJITBindings(PyObject* module) {
             docstring << "  " << op->schema() << "\n";
           }
 
-          return py::cpp_function(
-              [operations](py::args args, py::kwargs kwargs) {
+          // To support __torch_function__ override mechanism, `func` needs to
+          // provide a pointer, which points to the original method, i.e.,
+          // itself. py::cpp_function doesn't support specifying the function
+          // signature without real implementation, so we we use a trick here.
+          // First, we create a function holder pointing to an empty
+          // cpp_function. Then we use a reference as the alias of the empty
+          // function. Then we define the real returned function, and update
+          // the value of the function pointed by the holder using the alias.
+          // With this trick, we are able to reference self inside the returned
+          // function.
+          auto func_holder = std::make_unique<py::cpp_function>();
+          py::cpp_function& func_ref = *func_holder;
+          auto func = py::cpp_function(
+              [operations, symbol, self_ptr = std::move(func_holder)](
+                  py::args args, py::kwargs kwargs) {
+                std::vector<py::handle> overloaded_args;
```

Comment:
If I recall correctly, when @hameerabbasi was doing benchmarking, having overloaded argument detection separate from normal argument parsing was a noticeable overhead (that's why it's interleaved with regular argument parsing in the normal implementation). This implementation here is probably OK for now, but if you ever turn around and want to start optimizing the performance of `torch.ops`, this code here is going to make things harder to do so. But it is probably OK to not worry about this too much right now, because there are probably a bunch of performance problems with torch.ops today already, and so if we want to tackle perf here we should actually measure.

Regardless, it might be good to do a little microbenchmark and see exactly how much we're losing here.

cc @ljk53

Reply:
okay, then i will just use py::import to fetch the pointer to original method then.

Reply:
@ezyang do you know how we benchmarked the torch.ops before? also cc: @smessmer

Reply:
I don't think we've done any torch.ops benchmarking; c.f., https://github.com/pytorch/pytorch/pull/42341 . But I would go about benchmarking them the same way I benchmark normal ops. cc @robieta @ngimel @wconstab